src/backend.md

```diff
@@ -18,6 +19,236 @@
 Table of contents can be generated in services like http://ecotrust-canada.github.io/markdown-toc/
 -->
 
+## Project structure
+
+Project structure should follow next principles:
+
+1. Not needed to move files without changing their purpose.
+
+1. It's easy to identify a single place where a new file should be located.
+
+1. It's easy to understand where to look files/functionality up.
+
+### Top level basic project structure
+
+- src
+  - controllers
+  - events
+  - jobs
+  - lib
+  - models
+  - sdk
+  - services
+  - tests
+
+#### controllers directory
+
+1. Controllers for resources should be placed into separated directories named as resource. For example: `user`, `role`, e.t.c.
+2. Each resource directory should have 4 files. For example if resource is `user`:
+
+| File extension            | description                        |
+| ------------------------- | ---------------------------------- |
+| `user.controller.ts`      | controller handlers implementation |
+| `user.controller.test.ts` | tests for controller handlers      |
+| `user.endpoints.ts`       | endpoints declaration              |
+| `handlers.ts`             |                                    |
+
+`handlers.ts` - special file which contains only 1 export instruction:
+
+```ts
+export * from './user.controller';
+```
+
+Controllers directory structure example:
+
+```
+- controllers
+  - user
+    - user.controller.ts
+    - user.controller.test.ts
+    - user.endpoints.ts
+    - handlers.ts
+  - endpoints.ts
+  - index.ts
+```
+
+For some resources `*.controller.test.ts` file can become huge (> 1500 rows). So if it exceeds some comfortable size for reading it may be splitted into separate files and put them into `tests` directory inside resource directory.
+
+Assume that user resource has following endpoint names: `getUsers`, `getUser`, `createUser`. So each test file should be named as `{endpoint name}.controller.test.ts`.
+
+Example:
+
+```
+- controllers
+  - user
+    - tests
+      - getUsers.controller.test.ts
+      - getUser.controller.test.ts
+      - createUser.controller.test.ts
+    - user.controller.ts
+    - user.endpoints.ts
+    - handlers.ts
+  - endpoints.ts
+  - index.ts
+```
+
+#### events directory
+
+Event handlers for resources should be placed into separated directories named as resource. For example: `user`, `role`, e.t.c.
+
+Events directory structure example:
+
+```
+- events
+  - user
+    - user.event.ts
+    - user.event.test.ts
+```
+
+#### jobs directory
+
+Job handlers for resources should be placed into separated directories named as resource. For example: `user`, `role`, e.t.c.
+
+Jobs directory structure example:
+
+```
+- jobs
+  - user
+    - user.job.ts
+    - user.job.test.ts
+  - jobNames.ts
+  - jobHandlers.ts
```

Comment:
But currently, they live in `services`. Do you suggest to move to `jobs`?

Reply:
Good question.

In our projects we have completely different places for these files

First:
```
- jobs
 ...
  - jobNames.ts
  - jobHandlers.ts
```
Second:
```
- jobs
 ...
  - jobHandlers.ts
- services
  ...
  - jobNames.ts
```
Third:
```
- jobs
 ...
- services
  ...
  - jobHandlers.ts
  - jobNames.ts
```

Remember I made proposal to set up linter for control imports direction between our layers?
Third way is no good because it makes a lot of imports like `services -> jobs` which are not allowable.

Second option is much better because `jobHandlers.ts` has it own imports from jobs/... and `jobNames.ts` is on service level. So all possible places for enqueue jobs looks allowable:
1.  controllers -> `import JobNames.ts from ../services`
2. jobs -> `import JobNames.ts from ../services`
3. events -> `import JobNames.ts from ../services`
4. services -> `import JobNames.ts from ../services`

But... I think you have a great experience to make a new jobs and it is so useful to have `jobNames.ts` and `jobHandlers.ts` at the same place near all job handler files in `src/jobs` directory.

So personally I prefer to choose the first way and place these files into jobs folder.

Reply:
It is a strong argument to have an ability to restrict imports. So I would prefer to have it to to stay with (3).

Why (3) is better then (2):

- it is confusing to have these 2 files in different places

Reply:
Generally, I would like to add description of the status quo in this PR, and to discuss changes separately